extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqttbroker/internal/discovery/NetworkDiscoveryService.java

```diff
@@ -0,0 +1,231 @@
+     */
+    protected void scanTarget(MqttBrokerConnection testConnection)
+            throws ConfigurationException, MqttException, InterruptedException {
+        // Block wait for the callback
+        final Lock lock = new ReentrantLock();
+        lock.lock();
+
+        testConnection.addConnectionObserver((isConnected, error) -> lock.unlock());
+
+        // Start the connection and wait until timeout or connected callback returns.
+        testConnection.start();
+        lock.tryLock(TIMEOUT_IN_MS, TimeUnit.MILLISECONDS);
+
+        if (testConnection.connectionState() == MqttConnectionState.CONNECTED) {
+            logger.trace("Found service device at {}",
+                    testConnection.getHost() + ":" + String.valueOf(testConnection.getPort()));
+
+            Map<String, Object> properties = new HashMap<>();
+            properties.put(MqttBrokerBindingConstants.PARAM_BROKER_HOST, testConnection.getHost());
+            properties.put(MqttBrokerBindingConstants.PARAM_BROKER_PORT, testConnection.getPort());
+
+            ThingUID thingUID = MqttThingID.getThingUID(testConnection.getHost(), testConnection.getPort());
+            thingDiscovered(DiscoveryResultBuilder.create(thingUID).withTTL(120).withProperties(properties)
+                    .withLabel("Mqtt Broker (" + testConnection.getHost() + ")").build());
+        }
+
+        testConnection.stop();
+        testConnection = null;
+    }
+
+    /**
+     * Increase a counter and calls stopScan() if counter equals total.
+     *
+     * @param total The total size
+     */
+    synchronized protected void stopScanIfAllScanned(int total) {
+        scannedIPcount += 1;
+        if (scannedIPcount == total) {
+            logger.trace("Scan of {} IPs successful", scannedIPcount);
+            stopScan();
+        }
+    }
+
+    /**
+     * Return a list of IP v4 addresses to be scanned for the discovery
+     */
+    protected List<String> getScannableIPs() {
+        return NetUtil.getAllInterfaceAddresses().stream().filter(i -> i.getAddress() instanceof Inet4Address)
+                .map(interfaceIP -> Arrays
+                        .asList(new SubnetUtils(interfaceIP.getAddress().getHostAddress()).getInfo().getAllAddresses())
+                        .subList(0, MAX_IPS_PER_INTERFACE))
+                .flatMap(List::stream).collect(Collectors.toList());
+    }
+
+    /**
+     * Starts the DiscoveryThread for each IP on each interface on the network
+     */
+    @Override
+    protected void startScan() {
+        if (executorService != null) {
+            return;
+        }
+        removeOlderResults(getTimestampOfLastScan(), null);
+        logger.trace("Starting Discovery");
+
+        List<String> networkIPs = getScannableIPs();
+
+        // We perform two scans per IP (plain/secure)
+        executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
```

Comment:
Hm, shouldn't we use the scheduler of the discovery service (protected member of the abstract discovery service)?